### DISCUSSION TITLE
Injecting values to controller arguments

### DISCUSSION THREAD

**Discussion by GithubUser8080** [playframework/playframework#11426] [General]
Hi, we are trying to emulate some of Springs features like automatic dto (or other value) injection in controller methods, validation, and security filtering. 

E.g

`myControllerMethod(User currentUser, @Valid Product productDto);`

The only interface i can see that provides a mechanism for controller value injection is QueryStringBindable and PathBindable. However, user is not based on the query string or path, but a header token, and the dto comes from the body. Is there any plan to enhance this functionality?  Because at the moment you have to add the json parsing, dto validation, and authentication, like the example in every controller method. I would like to centralize this and not copy paste it tens of times. Would that be possible? Thanks

**Comment by tsuyoshizawa** [comment]
Authentication using PlayFramework can do using ActionFilter on Scala to avoid copy and paste. https://www.playframework.com/documentation/2.8.x/ScalaActionsComposition#Validating-requests

I don't use the Play Framework with Java, the use of custom actions may be a solution.
https://www.playframework.com/documentation/2.8.x/JavaActionsComposition#Defining-custom-action-annotations

**Reply by rbueker** [reply]
Check out https://www.pac4j.org/ if you want a security framework.

You would use the Play implementation: https://github.com/pac4j/play-pac4j

There are demos for both Java: https://github.com/pac4j/play-pac4j-java-demo  and scala: https://github.com/pac4j/play-pac4j-scala-demo

**Reply by GithubUser8080** [reply]
Indeed we can use custom actions, but you still need a **getSomethingFromRequest(request)** in every controller method for every possible injectable value. Play could easily provide a Bindable interface much like queryBindable that supports injection, and binding from body, to make the framework a bit more declarative. 

Something like

```
public class UserInjectable implementsControllerInjectable<User> {
     @Inject 
     public UserInjectable(...config..etc){
     } 

     public User get(Request request){... }
}
```

Then we can register this in a repository and inject User/Principal objects in controller methods, and do

`CompletionStage<Result> myControllerMethod(User user){ ... }`


Just an idea
